src/sync/channel.zig

```diff
@@ -166,105 +168,519 @@ pub fn Channel(comptime T: type) type {
+
+                    if (offset + 1 == BLOCK_CAP) {
+                        block.?.destroy(0, channel.allocator);
+                    } else if (slot.state.fetchOr(READ, .acq_rel) & DESTROY != 0) {
+                        block.?.destroy(offset + 1, channel.allocator);
+                    }
+
+                    return value;
+                }
+            }
+        }
+
+        pub fn len(channel: *Self) usize {
+            while (true) {
+                var tail = channel.tail.index.load(.seq_cst);
+                var head = channel.head.index.load(.seq_cst);
+
+                if (channel.tail.index.load(.seq_cst) == tail) {
+                    tail &= ~((@as(usize, 1) << SHIFT) - 1);
+                    head &= ~((@as(usize, 1) << SHIFT) - 1);
+
+                    if ((tail >> SHIFT) & (LAP - 1) == (LAP - 1)) {
+                        tail +%= (1 << SHIFT);
+                    }
+                    if ((head >> SHIFT) & (LAP - 1) == (LAP - 1)) {
+                        head +%= (1 << SHIFT);
+                    }
+
+                    const lap = (head >> SHIFT) / LAP;
+                    tail -%= (lap * LAP) << SHIFT;
+                    head -%= (lap * LAP) << SHIFT;
+
+                    tail >>= SHIFT;
+                    head >>= SHIFT;
+
+                    return tail - head - tail / LAP;
+                }
+            }
+        }
+
+        pub fn isEmpty(channel: *Self) bool {
+            const head = channel.head.index.load(.seq_cst);
+            const tail = channel.tail.index.load(.seq_cst);
+            return (head >> SHIFT) == (tail >> SHIFT);
+        }
+
+        pub fn deinit(channel: *Self) void {
+            var head = channel.head.index.raw;
+            var tail = channel.tail.index.raw;
+            var block = channel.head.block.raw;
+
+            head &= ~((@as(usize, 1) << SHIFT) - 1);
+            tail &= ~((@as(usize, 1) << SHIFT) - 1);
+
+            while (head != tail) {
+                const offset = (head >> SHIFT) % LAP;
+
+                if (offset >= BLOCK_CAP) {
+                    const next = block.?.next.raw;
+                    channel.allocator.destroy(block.?);
+                    block = next;
+                }
+
+                head +%= (1 << SHIFT);
+            }
+
+            if (block) |b| {
+                channel.allocator.destroy(b);
+            }
+        }
+
+        pub fn close(channel: *Self) void {
+            _ = channel;
+        }
+    };
+}
+
 const Block = struct {
     num: u32 = 333,
     valid: bool = true,
     data: [1024]u8 = undefined,
 };
 
-const BlockChannel = Channel(Block);
-const BlockPointerChannel = Channel(*Block);
-
 const logger = std.log.scoped(.sync_channel_tests);
 
-fn testReceiver(chan: *BlockChannel, recv_count: *Atomic(usize), id: u8) void {
-    _ = id;
-    while (chan.receive()) |v| {
-        _ = v;
-        _ = recv_count.fetchAdd(1, .seq_cst);
+fn testUsizeReceiver(chan: anytype, recv_count: usize) void {
```

Comment:
@dnut prev found some channel bugs with the test in this commit: https://github.com/Syndica/sig/commit/bac636ecacc5ac7f21df62d276f87a64f05fdad7

its pretty simple - probs worth it to port it over or have similar assertions to make sure everything works expected